doc/source/language/user_functions.rst

```diff
@@ -0,0 +1,339 @@
+DECLARE FUNCTION command, which has syntax as follows:
+
+  ``declare function`` *identifier* ``{`` *statements* ``}`` *optional dot (.)*
+
+example::
+
+    // Print the string you pass in, in one of the 4 corners
+    // of the terminal:
+    //   mode = 1 for upper-left, 2 for upper-right, 3
+    //          for lower-left, and 4 for lower-right:
+    //
+    declare function print_corner {
+      declare parameter mode.
+      declare parameter text.
+
+      declare row.
+      declare col.
+
+      if mode = 1 or mode = 3 {
+        set col to 0.
+      } else {
+        set col to terminal:width - text:length.
+      }.
+      if mode = 1 or mode = 2 {
+        set row to 0.
+      } else {
+        set row to terminal:height - 1.
+      }.
+
+      print text at (col, row).
+    }.
+
+    // An example of calling it:
+    
+    print_corner(4,"That's me in the corner").
+
+``Declare function`` can appear anywhere in a kerboscript program,
+and once its been "parsed" by the compiler, the function can be called
+from anywhere in the program.  
+
+The best design pattern is probably to create your library of function
+calls as one or more separate .ks files that contain ONLY function
+definitions and nothing else in them.  Then when you "run" the file
+containing the functions, what you're really doing is just loading
+the function definitions into memory so they can be called by other
+programs.  At the top of your main script you can then "run" the
+other scripts containing the library of functions to get them
+compiled into memory.
+
+``DECLARE PARAMETER``
+---------------------
+
+If your function expects to have parameters passed into it, you can
+use the :ref:`DECLARE PARAMETER <declare parameter>` command to do
+so.  This is the same command as is used to declare parameters for
+running a whole script.  By putting a DECLARE PARAMETER statement
+inside a function, you tell the kerboscript compiler that you want
+the parameter to be for that function, not for the whole script.
+
+An example of using ``declare parameter`` can be seen in the example
+above, where it is used for the ``mode`` and ``text`` parameters.
+
+``DECLARE``
+-----------
+
+(aka: **local variables**)
+
+If your function needs to make a local variable, it can do so using
+the :ref:`DECLARE <declare>` command.  Whenever the DECLARE command is
+seen inside a function, the compiler assumes the variable is meant to
+be local to that function's block.  This also works with recursion.
+If you recursively call a function again and again, there will be 
+new copies stacked up of all the local variables made with DECLARE,
+but not of the variables implicitly made global without DECLARE.
+
+An example of using ``declare`` for a local variable can be seen in
+the example above, where it is used for the ``row`` and ``col`` variables.
+
+A more in-depth explanation of kerboscript's scoping rules and how they
+work is found :ref:`on another page <scope>`
+
+``DECLARE TO``
```

Comment:
declare to has the benefit of being consistent with SET .. TO 

If the grammar bothers you, we could do DECLARE .. WITH?

eg DECLARE foo with 'bar'.